modules/core/src/main/java/org/apache/ignite/internal/util/VersatileReadWriteLock.java

```diff
@@ -0,0 +1,498 @@
+ *
+ * <p>Write lock acquire requests are prioritized over read lock acquire requests. That is, if both read and write lock
+ * acquire requests are received when the write lock is held by someone else, then, on its release, the write lock attempt will be served
+ * first.
+ *
+ * <p>Lock owners are not tracked.
+ *
+ * <p>Asynchronous locking methods may complete the futures either in the calling thread (if they were able to immediately acquire
+ * the requested lock) or in the supplied pool (if they had to wait for a release to happen before being able to satisfy the request).
+ *
+ * <p>Asynchronous locking methods never use spin loops. They do use CAS loops, but these are mostly very short.
+ */
+public class VersatileReadWriteLock {
+    /** The highest bit of the state signals that the write lock is acquired. */
+    private static final int WRITE_LOCK_BITS = 1 << 31;
+
+    /** 31 lower bits represent a counter of read locks that qre acquired. */
+    private static final int READ_LOCK_BITS = ~WRITE_LOCK_BITS;
+
+    /**
+     * State 0 means that both read and write locks are available for acquiring (as no locks are acquired).
+     *
+     * @see #state
+     */
+    private static final int AVAILABLE = 0;
+
+    /** How much time to sleep on each iteration of a spin loop (milliseconds). */
+    private static final int SLEEP_MILLIS = 10;
+
+    /** {@link VarHandle} used to access the {@code pendingWLocks} field. */
+    private static final VarHandle PENDING_WLOCKS_VH;
+
+    /** {@link VarHandle} used to access the {@code state} field. */
+    private static final VarHandle STATE_VH;
+
+    static {
+        try {
+            STATE_VH = MethodHandles.lookup()
+                .findVarHandle(VersatileReadWriteLock.class, "state", int.class);
+
+            PENDING_WLOCKS_VH = MethodHandles.lookup()
+                .findVarHandle(VersatileReadWriteLock.class, "pendingWriteLocks", int.class);
+        } catch (ReflectiveOperationException e) {
+            throw new ExceptionInInitializerError(e);
+        }
+    }
+
+    /**
+     * Main state of the lock.
+     * <ul>
+     *     <li>The highest bit is for whether the write lock is acquired or not.</li>
+     *     <li>The remaining bits store a counter representing the acquired read locks.</li>
+     * </ul>
+     */
+    @SuppressWarnings("unused")
+    private volatile int state;
+
+    /**
+     * Number of pending write attempts to acquire the write lock. It is used to prioritize write lock attempts over read
+     * lock attempts when the write lock has been released (so, if both an attempt to acquire the write lock and an attempt to acquire the
+     * read lock are waiting for write lock to be released, a write lock attempt will be served first when the release happens).
+     */
+    @SuppressWarnings("unused")
+    private volatile int pendingWriteLocks;
```

Comment:
Why not making both parts a single long value?

Reply:
Are you asking about `state` and `pendingWriteChecks`? They don't always get CASed together, so I don't see why we would want to merge them together.

Reply:
They're on a single cache line anyway, so there's already a naturally occurring contention

Reply:
If they are in the same long, they are guaranteed to be on the same cache line; if they aren't, maybe they will be in different cache lines (this is possible at least in theory, right?). So no gain here, maybe even a loss.

Also, squashing them together complicates the code, which is also a loss.

I see no gain from squashing, only losses.